File: app/src/main/java/com/clockmods/weather/QWeatherClient.java
package com.clockmods.weather;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Locale;

import javax.net.ssl.HttpsURLConnection;

import static com.clockmods.weather.WeatherModels.WeatherDetail;
import static com.clockmods.weather.WeatherModels.WeatherDisplayData;

public final class QWeatherClient {
    private final String host;
    private final int timeoutMs;
    private final TlsSocketFactory socketFactory;

    public QWeatherClient(Context context, String host) { this(context, host, 15000); }
    public QWeatherClient(Context context, String host, int timeoutMs) {
        this.host = host; this.timeoutMs = timeoutMs;
        this.socketFactory = TlsSocketFactory.create(context.getApplicationContext());
    }

    public WeatherDisplayData fetch(double latitude, double longitude) throws Exception {
        return fetch(latitude, longitude, false);
    }

    public WeatherDisplayData fetch(double latitude, double longitude, boolean detailed) throws Exception {
        JSONObject location = getLocation(latitude, longitude);
        double lat = optDouble(location, "lat", latitude);
        double lon = optDouble(location, "lon", longitude);
        return fetchNow(location.getString("id"), location.optString("adm2", ""),
                location.optString("name", ""), lat, lon, detailed);
    }

    public WeatherDisplayData fetchLocation(String locationId, String city, String district) throws Exception {
        return fetchNow(locationId, city, district, Double.NaN, Double.NaN, false);
    }

    public WeatherDisplayData fetchLocation(String locationId, String city, String district,
            double latitude, double longitude, boolean detailed) throws Exception {
        return fetchNow(locationId, city, district, latitude, longitude, detailed);
    }

    private WeatherDisplayData fetchNow(String locationId, String city, String district,
            double latitude, double longitude, boolean detailed) throws Exception {
        String path = "/v7/weather/now?location=" + urlEncode(locationId) + "&lang=zh";
        JSONObject body = request(path);
        JSONObject now = body.getJSONObject("now");
        WeatherDetail detail = detailed
                ? buildDetail(now, latitude, longitude) : null;
        return new WeatherDisplayData(locationId, city, district,
                now.optString("text", ""), now.optString("icon", ""),
                now.optString("temp", "--"), System.currentTimeMillis(), detail);
    }

    private WeatherDetail buildDetail(JSONObject now, double latitude, double longitude) {
        String warning = null;
        String aqiValue = null;
        String aqiCategory = null;
        if (!Double.isNaN(latitude) && !Double.isNaN(longitude)) {
            warning = fetchWarning(latitude, longitude);
            String[] aqi = fetchAirQuality(latitude, longitude);
            aqiValue = aqi[0];
            aqiCategory = aqi[1];
        }
        return new WeatherDetail(now.optString("feelsLike", ""), now.optString("humidity", ""),
                now.optString("windDir", ""), now.optString("windScale", ""),
                now.optString("precip", ""), warning, aqiValue, aqiCategory);
    }

    private String fetchWarning(double latitude, double longitude) {
        try {
            JSONObject body = requestRaw("/weatheralert/v1/current/"
                    + formatCoordinate(latitude) + "/" + formatCoordinate(longitude) + "?lang=zh");
            JSONArray alerts = body.optJSONArray("alerts");
            if (alerts == null || alerts.length() == 0) return null;
            return formatWarning(alerts.getJSONObject(0));
        } catch (Exception ignored) { return null; }
    }

    static String formatWarning(JSONObject alert) {
        String headline = alert.optString("headline", "").trim();
        if (headline.length() > 0) return headline;

        JSONObject eventType = alert.optJSONObject("eventType");
        String event = eventType == null ? "" : eventType.optString("name", "").trim();
        if (event.length() == 0) return null;

        JSONObject color = alert.optJSONObject("color");
        String colorCode = color == null ? "" : color.optString("code", "").trim();
        String colorName = warningColorName(colorCode);
        return event + colorName + "预警";
    }

    private static String warningColorName(String code) {
        if (code.length() == 0) return "";
        switch (code.toLowerCase(Locale.US)) {
            case "white": return "白色";
            case "gray": return "灰色";
            case "green": return "绿色";
            case "blue": return "蓝色";
            case "yellow": return "黄色";
            case "amber": return "琥珀色";
            case "orange": return "橙色";
            case "red": return "红色";
            case "purple": return "紫色";
            case "black": return "黑色";
            default: return code;
        }
    }

    private String[] fetchAirQuality(double latitude, double longitude) {
        try {
            JSONObject body = requestRaw("/airquality/v1/current/"
                    + formatCoordinate(latitude) + "/" + formatCoordinate(longitude) + "?lang=zh");
            JSONArray indexes = body.optJSONArray("indexes");
            if (indexes == null || indexes.length() == 0) return new String[] {null, null};
            JSONObject index = indexes.getJSONObject(0);
            for (int i = 0; i < indexes.length(); i++) {
                if ("qaqi".equals(indexes.getJSONObject(i).optString("code"))) {
                    index = indexes.getJSONObject(i);
                    break;
                }
            }
            String value = index.optString("aqiDisplay", "");
            String category = index.optString("category", "");
            return new String[] {value.length() > 0 ? value : null,
                    category.length() > 0 ? category : null};
        } catch (Exception ignored) { return new String[] {null, null}; }
    }

    private JSONObject getLocation(double latitude, double longitude) throws Exception {
        JSONObject body = request("/geo/v2/city/lookup?location=" + formatLocation(latitude, longitude)
                + "&range=cn&number=1&lang=zh");
        JSONArray locations = body.getJSONArray("location");
        if (locations.length() == 0) throw new IOException("No QWeather location");
        return locations.getJSONObject(0);
    }

    static String formatLocation(double latitude, double longitude) {
        return String.format(Locale.US, "%.2f,%.2f", longitude, latitude);
    }

    static String formatCoordinate(double value) {
        return String.format(Locale.US, "%.2f", value);
    }

    private static double optDouble(JSONObject object, String key, double fallback) {
        String value = object.optString(key, "");
        try { return value.length() > 0 ? Double.parseDouble(value) : fallback; }
        catch (NumberFormatException e) { return fallback; }
    }

    private JSONObject request(String path) throws Exception {
        JSONObject json = requestRaw(path);
        if (!"200".equals(json.optString("code"))) {
            throw new IOException("QWeather error: " + json.optString("code"));
        }
        return json;
    }

    private JSONObject requestRaw(String path) throws Exception {
        HttpURLConnection connection = (HttpURLConnection) new URL("https://" + host + path).openConnection();
        try {
            if (connection instanceof HttpsURLConnection && socketFactory != null) {
                ((HttpsURLConnection) connection).setSSLSocketFactory(socketFactory);
            }
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(timeoutMs);
            connection.setReadTimeout(timeoutMs);
            connection.setRequestProperty("Authorization", "Bearer " + QWeatherSigner.token(
                    QWeatherConfig.credentialId(), QWeatherConfig.projectId(),
                    QWeatherConfig.privateKeyBase64(), System.currentTimeMillis() / 1000L));
            int status = connection.getResponseCode();
            String response = read(status >= 200 && status < 300
                    ? connection.getInputStream() : connection.getErrorStream());
            JSONObject json = new JSONObject(response);
            if (status < 200 || status >= 300) {
                throw new IOException("QWeather error: " + json.optString("code", String.valueOf(status)));
            }
            return json;
        } finally {
            connection.disconnect();
        }
    }

    private static String read(InputStream input) throws IOException {
        if (input == null) throw new IOException("Empty QWeather response");
        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(input, "UTF-8"))) {
            String line; while ((line = reader.readLine()) != null) output.append(line);
        }
        return output.toString();
    }

    private static String urlEncode(String value) throws Exception {
        return java.net.URLEncoder.encode(value, "UTF-8");
    }
}

File: app/src/main/java/com/clockmods/weather/QWeatherClient.kt
package com.clockmods.weather

import android.content.Context
import com.clockmods.background.ClockPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.Locale
import javax.net.ssl.HttpsURLConnection

class QWeatherClient @JvmOverloads constructor(context: Context, private val host: String, private val timeoutMs: Int = 15000) {
    private val socketFactory = TlsSocketFactory.create(context.applicationContext)
    private val lang = apiLang(ClockPreferences(context).getClockLanguage())
    @Throws(Exception::class) fun fetch(latitude: Double, longitude: Double) = fetch(latitude, longitude, false)
    @Throws(Exception::class) fun fetch(latitude: Double, longitude: Double, detailed: Boolean): WeatherModels.WeatherDisplayData { val location = getLocation(latitude, longitude); return fetchNow(location.getString("id"), location.optString("adm2", ""), location.optString("name", ""), optDouble(location, "lat", latitude), optDouble(location, "lon", longitude), detailed) }
    @Throws(Exception::class) fun fetchLocation(locationId: String, city: String, district: String) = fetchNow(locationId, city, district, Double.NaN, Double.NaN, false)
    @Throws(Exception::class) fun fetchLocation(locationId: String, city: String, district: String, latitude: Double, longitude: Double, detailed: Boolean) = fetchNow(locationId, city, district, latitude, longitude, detailed)
    @Throws(Exception::class) fun fetchDaily(latitude: Double, longitude: Double): WeatherModels.DailyForecastData { val location = getLocation(latitude, longitude); return fetchDailyLocation(location.getString("id"), location.optString("adm2", ""), location.optString("name", "")) }
    @Throws(Exception::class) fun fetchDailyLocation(locationId: String, city: String, district: String) = parseDailyForecast(request("/v7/weather/3d?location=${urlEncode(locationId)}&lang=$lang"), locationId, city, district, System.currentTimeMillis())
    private fun fetchNow(locationId: String, city: String, district: String, latitude: Double, longitude: Double, detailed: Boolean): WeatherModels.WeatherDisplayData { val now = request("/v7/weather/now?location=${urlEncode(locationId)}&lang=$lang").getJSONObject("now"); return WeatherModels.WeatherDisplayData(locationId, city, district, now.optString("text", ""), now.optString("icon", ""), now.optString("temp", "--"), System.currentTimeMillis(), if (detailed) buildDetail(now, latitude, longitude) else null) }
    private fun buildDetail(now: JSONObject, latitude: Double, longitude: Double): WeatherModels.WeatherDetail { val warning = if (!latitude.isNaN() && !longitude.isNaN()) fetchWarning(latitude, longitude) else null; val aqi: Array<String?> = if (!latitude.isNaN() && !longitude.isNaN()) fetchAirQuality(latitude, longitude) else arrayOfNulls(2); return WeatherModels.WeatherDetail(now.optString("feelsLike", ""), now.optString("humidity", ""), now.optString("windDir", ""), now.optString("windScale", ""), now.optString("precip", ""), warning, aqi[0], aqi[1]) }
    private fun fetchWarning(latitude: Double, longitude: Double): String? = try { formatWarnings(requestRaw("/weatheralert/v1/current/${formatCoordinate(latitude)}/${formatCoordinate(longitude)}?lang=$lang").optJSONArray("alerts"), lang) } catch (_: Exception) { null }
    private fun fetchAirQuality(latitude: Double, longitude: Double): Array<String?> {
        return try {
            val indexes = requestRaw("/airquality/v1/current/${formatCoordinate(latitude)}/${formatCoordinate(longitude)}?lang=$lang").optJSONArray("indexes") ?: return arrayOfNulls(2)
            if (indexes.length() == 0) return arrayOfNulls(2)
            var index = indexes.getJSONObject(0)
            for (i in 0 until indexes.length()) if (indexes.getJSONObject(i).optString("code") == "qaqi") { index = indexes.getJSONObject(i); break }
            arrayOf(index.optString("aqiDisplay", "").takeIf(String::isNotEmpty), index.optString("category", "").takeIf(String::isNotEmpty))
        } catch (_: Exception) { arrayOfNulls(2) }
    }
    private fun getLocation(latitude: Double, longitude: Double): JSONObject { val locations = request("/geo/v2/city/lookup?location=${formatLocation(latitude, longitude)}&range=cn&number=1&lang=$lang").getJSONArray("location"); if (locations.length() == 0) throw IOException("No QWeather location"); return locations.getJSONObject(0) }
    private fun request(path: String): JSONObject = requestRaw(path).also { if (it.optString("code") != "200") throw IOException("QWeather error: ${it.optString("code")}") }
    private fun requestRaw(path: String): JSONObject { val connection = URL("https://$host$path").openConnection() as HttpURLConnection; try { if (connection is HttpsURLConnection && socketFactory != null) connection.sslSocketFactory = socketFactory; connection.requestMethod = "GET"; connection.connectTimeout = timeoutMs; connection.readTimeout = timeoutMs; connection.setRequestProperty("Authorization", "Bearer " + QWeatherSigner.token(QWeatherConfig.credentialId(), QWeatherConfig.developerId(), QWeatherConfig.projectId(), QWeatherConfig.privateKeyBase64(), System.currentTimeMillis() / 1000L)); val status = connection.responseCode; val response = (if (status in 200..299) connection.inputStream else connection.errorStream)?.bufferedReader(Charsets.UTF_8)?.use { it.readText() } ?: throw IOException("Empty QWeather response"); val json = JSONObject(response); if (status !in 200..299) throw IOException("QWeather error: ${json.optString("code", status.toString())}"); return json } finally { connection.disconnect() } }

    companion object {
        @JvmStatic fun apiLang(clockLanguage: String?): String = when (clockLanguage) { ClockPreferences.LANGUAGE_ENGLISH -> "en"; ClockPreferences.LANGUAGE_TRADITIONAL -> "zh-hant"; else -> "zh" }
        @JvmStatic @Throws(Exception::class) fun parseDailyForecast(body: JSONObject, locationId: String, city: String, district: String, updatedAt: Long): WeatherModels.DailyForecastData { val daily = body.optJSONArray("daily") ?: throw IOException("Empty daily forecast"); val entries = ArrayList<WeatherModels.DailyForecast>(); for (index in 0 until daily.length()) { val day = daily.optJSONObject(index) ?: continue; val date = day.optString("fxDate", "").trim(); if (date.isNotEmpty()) entries += WeatherModels.DailyForecast(date, day.optString("tempMin", ""), day.optString("tempMax", ""), day.optString("iconDay", ""), day.optString("textDay", ""), day.optString("windDirDay", ""), day.optString("windScaleDay", ""), day.optString("humidity", "")) }; if (entries.isEmpty()) throw IOException("Empty daily forecast"); return WeatherModels.DailyForecastData(locationId, city, district, updatedAt, entries) }
        @JvmStatic fun formatWarnings(alerts: JSONArray?): String? = formatWarnings(alerts, "zh")
        @JvmStatic fun formatWarnings(alerts: JSONArray?, lang: String): String? { if (alerts == null || alerts.length() == 0) return null; val values = ArrayList<String>(); for (index in 0 until minOf(alerts.length(), 20)) alerts.optJSONObject(index)?.let { formatWarning(it, lang)?.takeIf(String::isNotEmpty)?.let(values::add) }; return values.joinToString("\n").takeIf(String::isNotEmpty) }
        @JvmStatic fun formatWarning(alert: JSONObject): String? = formatWarning(alert, "zh")
        @JvmStatic fun formatWarning(alert: JSONObject, lang: String): String? { alert.optString("headline", "").trim().takeIf(String::isNotEmpty)?.let { return it }; val event = alert.optJSONObject("eventType")?.optString("name", "")?.trim().orEmpty(); if (event.isEmpty()) return null; val code = alert.optJSONObject("color")?.optString("code", "")?.trim().orEmpty(); if (lang == "en") return event + warningColorEnglish(code).let { if (it.isEmpty()) " Warning" else " $it Warning" }; val traditional = lang == "zh-hant"; return event + warningColorChinese(code, traditional) + if (traditional) "預警" else "预警" }
        private fun warningColorEnglish(code: String): String = when (code.lowercase(Locale.US)) { "white" -> "White"; "gray" -> "Gray"; "green" -> "Green"; "blue" -> "Blue"; "yellow" -> "Yellow"; "amber" -> "Amber"; "orange" -> "Orange"; "red" -> "Red"; "purple" -> "Purple"; "black" -> "Black"; else -> code }
        private fun warningColorChinese(code: String, traditional: Boolean): String { val common = mapOf("white" to "白色", "gray" to "灰色", "amber" to "琥珀色", "orange" to "橙色", "black" to "黑色"); val simplified = mapOf("green" to "绿色", "blue" to "蓝色", "yellow" to "黄色", "red" to "红色", "purple" to "紫色"); val traditionalMap = mapOf("green" to "綠色", "blue" to "藍色", "yellow" to "黃色", "red" to "紅色", "purple" to "紫色"); val lower = code.lowercase(Locale.US); return common[lower] ?: (if (traditional) traditionalMap[lower] else simplified[lower]) ?: code }
        @JvmStatic fun formatLocation(latitude: Double, longitude: Double) = String.format(Locale.US, "%.2f,%.2f", longitude, latitude)
        @JvmStatic fun formatCoordinate(value: Double) = String.format(Locale.US, "%.2f", value)
        private fun optDouble(objectValue: JSONObject, key: String, fallback: Double) = objectValue.optString(key, "").toDoubleOrNull() ?: fallback
        private fun urlEncode(value: String) = URLEncoder.encode(value, "UTF-8")
    }
}
